### PR TITLE
[Core/Observability 1/N] Add a "running" state to task status

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR adds 2 more states into `TaskStatus`

```
enum TaskStatus {
  // The task is scheduled properly and waiting for execution.
  // It includes time to deliver the task to the remote worker + queueing time
  // from the execution side.
  WAITING_FOR_EXECUTION = 5;
  // The task that is running.
  RUNNING = 6;
}
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/common.proto [hunk 1]
```diff
@@ -528,6 +528,10 @@ enum TaskStatus {
   SCHEDULED = 2;
   // The task finished successfully.
   FINISHED = 3;
+  // The task is submitted to a worker process to execute.
+  SUBMITTED_TO_WORKER = 4;
+  // The task has failed to be scheduled and being rescheduled.
+  RESCHEDULED = 5;
```
Comment (rkooo567) [inline src/ray/protobuf/common.proto:534]: Q: Maybe it is unnecessary, and we can reuse SCHEDULED state here.

**Comment by rkooo567** [comment]
Btw, one question I have is if "SUBMITTED_TO_WORKER" necessary. I originally wanted to implement "RUNNING", and I realized the current implementation cannot actually become "RUNNING" because for things like actor tasks, we have the execution side queuing (which means we can know the exact state only from the execution side, not a caller (owner) side).

This is the full picture

```
NIL
WAITING_FOR_DEPENDENCY
SCHEDULING - Detailed states: WAITING_FOR_WORKER,. WAITING_FOR_RUNTIME_ENV, WAITING_FOR_OBJECT_PULL, etc.
SUBMITTED_TO_WORKER # Is it necessary?
QUEUED_FOR_EXECUTION # on the execution side
RUNNING
FINISHED
```

File: src/ray/core_worker/task_manager.cc [hunk 2]
```diff
@@ -365,6 +383,7 @@ bool TaskManager::RetryTaskIfPossible(const TaskID &task_id) {
     } else {
       RAY_CHECK(num_retries_left == 0 || num_retries_left == -1);
     }
+    it->second.status = rpc::TaskStatus::RESCHEDULED;
```
Comment (scv119) [inline src/ray/core_worker/task_manager.cc:386]: this is not true is no retries left right?

File: src/ray/protobuf/common.proto (hunk 1 above)
Comment (scv119) [reply src/ray/protobuf/common.proto:534]: lgtm

File: src/ray/core_worker/task_manager.cc [hunk 3]
```diff
@@ -640,6 +659,16 @@ void TaskManager::MarkDependenciesResolved(const TaskID &task_id) {
   }
 }
 
+void TaskManager::MarkTaskSubmitted(const TaskID &task_id) {
+  absl::MutexLock lock(&mu_);
+  auto it = submissible_tasks_.find(task_id);
+  RAY_CHECK(it != submissible_tasks_.end())
```
Comment (scv119) [inline src/ray/core_worker/task_manager.cc:665]: hmm RAY_CHECK seems a bit extreme. Could we return bool and return early instead?

File: src/ray/protobuf/node_manager.proto [hunk 4]
```diff
@@ -300,7 +300,11 @@ message GetGcsServerAddressReply {
 message GetTasksInfoRequest {}
 
 message GetTasksInfoReply {
-  repeated TaskInfoEntry task_info_entries = 1;
+  repeated TaskInfoEntry owned_task_info_entries = 1;
+  // A list of task that's currently running.
+  // All tasks must be treated as TaskStatus::RUNNING
+  // from the caller side.
+  repeated bytes running_task_ids = 3;
```
Comment (scv119) [inline src/ray/protobuf/node_manager.proto:307]: nit should the index be 2 instead of 3?

**Review by scv119 — APPROVED** [review]

**Comment by rkooo567** [comment]
We need a bit more work on `ray memory` to display RUNNING state. Until we finish this task,  WAITING_FOR_EXECUTION will be a superset of `RUNNING` state when `ray memory` is used. Issue is tracked here https://github.com/ray-project/ray/issues/24835

**Comment by rkooo567** [comment]
Failed test should be unrelated

**Pull request merged**